src/Umbraco.Web/Compose/NestedContentPropertyComponent.cs

```diff
@@ -0,0 +1,128 @@
+using Umbraco.Core.Services.Implement;
+using Umbraco.Web.PropertyEditors;
+
+namespace Umbraco.Web.Compose
+{
+    public class NestedContentPropertyComponent : IComponent
+    {
+        public void Initialize()
+        {
+            ContentService.Copying += ContentService_Copying;
+            ContentService.Saving += ContentService_Saving;
+        }
+
+        private void ContentService_Copying(IContentService sender, CopyEventArgs<IContent> e)
+        {
+            // When a content node contains nested content property
+            // Check if the copied node contains a nested content
+            var nestedContentProps = e.Copy.Properties.Where(x => x.PropertyType.PropertyEditorAlias == Constants.PropertyEditors.Aliases.NestedContent);
+            UpdateNestedContentProperties(nestedContentProps, false);
+        }
+
+        private void ContentService_Saving(IContentService sender, ContentSavingEventArgs e)
+        {
+            // One or more content nodes could be saved in a bulk publish
+            foreach (var entity in e.SavedEntities)
+            {
+                // When a content node contains nested content property
+                // Check if the copied node contains a nested content
+                var nestedContentProps = entity.Properties.Where(x => x.PropertyType.PropertyEditorAlias == Constants.PropertyEditors.Aliases.NestedContent);
+                UpdateNestedContentProperties(nestedContentProps, true);
+            }
+        }
+
+        public void Terminate()
+        {
+            ContentService.Copying -= ContentService_Copying;
+            ContentService.Saving -= ContentService_Saving;
+        }
+
+        private void UpdateNestedContentProperties(IEnumerable<Property> nestedContentProps, bool onlyMissingKeys)
+        {
+            // Each NC Property on a doctype
+            foreach (var nestedContentProp in nestedContentProps)
+            {
+                // A NC Prop may have one or more values due to cultures
+                var propVals = nestedContentProp.Values;
+                foreach (var cultureVal in propVals)
+                {
+                    // Remove keys from published value & any nested NC's
+                    var updatedPublishedVal = CreateNestedContentKeys(cultureVal.PublishedValue?.ToString(), onlyMissingKeys);
+                    cultureVal.PublishedValue = updatedPublishedVal;
+
+                    // Remove keys from edited/draft value & any nested NC's
+                    var updatedEditedVal = CreateNestedContentKeys(cultureVal.EditedValue?.ToString(), onlyMissingKeys);
+                    cultureVal.EditedValue = updatedEditedVal;
+                }
+            }
+        }
+
+        private string CreateNestedContentKeys(string rawJson, bool onlyMissingKeys)
+        {
+            if (string.IsNullOrWhiteSpace(rawJson))
+                return rawJson;
+
+            // Parse JSON
+            var ncJson = JToken.Parse(rawJson);
```

Comment:
This should be even more safe, you don't really know if this string is json or if it is going to parse. You can at least use our string extension DetectIsJson to check if it might be json first